### DISCUSSION TITLE
Add support for large Bitmap aggregation (for additive, exact Distinct Counts)

### DISCUSSION THREAD

**Discussion by prmoore77** [duckdb/duckdb#3946] [Ideas]
This is to request a new SQL Aggregate function set to support Count Distinct via Bitmap operations.

Using a Bitmap technique to preserve exact sketches for Count Distinct aggregates is a technique that both [Oracle](https://danischnider.wordpress.com/2019/04/20/bitmap-based-countdistinct-functions-in-oracle-19c/) and [Snowflake](https://docs.snowflake.com/en/user-guide/querying-bitmaps-for-distinct-counts.html) have added.

This technique allows one to pre-aggregate distinct counts, which can then be used for further aggregation without the need for summary aggregates to use the granular data that was used to build the intermediate aggregates.

This can dramatically improve performance for hierarchical aggregation of Count Distinct measures by avoiding querying of granular data.

Specifically - these functions should be added to support Bitmap-based Count Distinct capabilities:
* [BITMAP_CONSTRUCT_AGG](https://docs.snowflake.com/en/sql-reference/functions/bitmap_construct_agg.html#bitmap-construct-agg)
* [BITMAP_BUCKET_NUMBER](https://docs.snowflake.com/en/sql-reference/functions/bitmap_bucket_number.html)
* [BITMAP_BIT_POSITION](https://docs.snowflake.com/en/sql-reference/functions/bitmap_bit_position.html#bitmap-bit-position)
* [BITMAP_OR_AGG](https://docs.snowflake.com/en/sql-reference/functions/bitmap_or_agg.html)
* [BITMAP_COUNT](https://docs.snowflake.com/en/sql-reference/functions/bitmap_count.html)

Currently - DuckDB does support BIT_OR aggregation on numeric data types (up to 16 Bytes via: HUGEINT) - but that only provides the ability to use 128 bits.  Snowflake can support up to 4k per bitmap, allowing much more compaction of distinct members.

**Comment by prmoore77** [comment]
Added PR: https://github.com/duckdb/duckdb/pull/4738 to start this work piece-meal...

**Comment by prmoore77** [comment]
I got this PostgreSQL example working to demonstrate the concept:
https://github.com/prmoore77/run-postgres-local/blob/main/sql/tpch_bitmap_aggregation_queries.sql

**Comment by prmoore77** [comment]
I believe the Snowflake actually used something like Boost's [dynamic bitset](https://www.boost.org/doc/libs/1_49_0/libs/dynamic_bitset/dynamic_bitset.html) C++ library for their specific implementation.  Using a bitset is much more efficient when the bitmap is sparse (many 0's vs. 1's).  For example - a 1000-bit bitmap with only 2 bits "on" would waste over 900 bits of information vs. a bitset.

**Reply by Mytherin** [reply]
[Roaring bitmaps](https://github.com/RoaringBitmap/CRoaring) seem like what you want to use for this.

**Reply by prmoore77** [reply]
I like that idea a lot - it seems they have (de)serialization methods that would make it relatively straightforward (for talented C++ devs, unlike myself :)).

**Reply by missinglink** [reply]
FYI: https://www.elastic.co/cn/blog/frame-of-reference-and-roaring-bitmaps

**Reply by prmoore77** [reply]
hi @missinglink - Bitmap aggregation was added to DuckDB with this PR: https://github.com/duckdb/duckdb/pull/6417 - it works very well...

**Reply by missinglink** [reply]
Cool, although I was hoping there was a way of constructing sparse/compressed bitmasks and convert to/from integers, this doesn't seem simple.
